### PR TITLE
GH-37917: [Parquet] Add OpenAsync for FileSource

### LINKED ISSUE #37917: [Parquet] Implement OpenAsync for FileSource
### Describe the enhancement requested

As discussed [here](https://github.com/apache/arrow/pull/37868), Parquet reader uses a synchronous FileSource::Open when starting to read a file: https://github.com/apache/arrow/blob/e038498c70207df1ac64b1aa276a5fd5e3cd306b/cpp/src/arrow/dataset/file_parquet.cc#L482

In the case of reading from S3, this Open method makes a HEAD request, which delays the start of processing of all other files when reading from a dataset.

Here are two illustrative images of the performance difference, reading a FileSystemDataset of 16 files with ~200K rows in total, in Pyarrow, with maximum fragment_readahead, pre_buffer=True. Files are on the y-axis and time is on the x-axis, threads are colored. Each point represents one request (HEAD or GET) to S3.

Here's the current behavior, where the first request for each file is processed on the blue thread:
Screenshot 2023-09-27 at 23 05 06

And here's the behavior with a WIP implementation of OpenAsync (note different x-axis scaling)
Screenshot 2023-09-27 at 23 04 49

In both images the first two (blue) points are from a separate request for one file to get the schema. It's still a bit of a mystery to me why in the async case the concurrent requests start only after the fourth request, seems like there could be some performance to be gained there as well.

### Component(s)

Parquet

### PR DESCRIPTION
**Pull request by eeroel**
### Rationale for this change

Improves performance of file reads with an expensive Open operation.

### Are there any user-facing changes?
No

* Closes: #37917

### CODE REVIEW

**Comment by eeroel** [comment]
Not sure about the details of the implementation but got it to compile and it seems to work as expected.

**Comment by mapleFU** [comment]
Windows failed is not related, but I think here we need a `clang-format`:

```
--- /arrow/cpp/src/arrow/dataset/file_parquet.cc
+++ /arrow/cpp/src/arrow/dataset/file_parquet.cc (after clang format)
```

you can just `clang-format -i /path/to/arrow/cpp/src/arrow/dataset/file_parquet.cc`

File: cpp/src/arrow/dataset/file_base.cc
Comment (mapleFU) [inline cpp/src/arrow/dataset/file_base.cc:93]: Oh, since `custom_open_` might still blocking here?

File: cpp/src/arrow/dataset/file_parquet.cc
Comment (mapleFU) [inline cpp/src/arrow/dataset/file_parquet.cc:499]: Would `WrapSourceError` called twice since L516-L519 already called that?

**Review by mapleFU — COMMENTED** [review]
General LGTM

File: cpp/src/arrow/dataset/file_base.cc
Comment (eeroel) [reply cpp/src/arrow/dataset/file_base.cc:93]: Yes, wasn't sure how to deal with it. FileSource would need its own `IOContext` instance to run it async?

File: cpp/src/arrow/dataset/file_parquet.cc
Comment (eeroel) [reply cpp/src/arrow/dataset/file_parquet.cc]: Good point, I removed the callback

File: cpp/src/arrow/dataset/file_base.cc
Comment (mapleFU) [reply cpp/src/arrow/dataset/file_base.cc:93]: You can leave a comment here, and keep it here? I think at lease, open via fileSystem could be help, and `custom_open_` won't be worse

Comment (eeroel) [reply cpp/src/arrow/dataset/file_base.cc:94]: 👍 left a todo

**Review by mapleFU — APPROVED** [review]

File: cpp/src/arrow/dataset/file_parquet.cc
Comment (mapleFU) [inline cpp/src/arrow/dataset/file_parquet.cc:494]: Can we just return `rfut` here?

Comment (eeroel) [reply cpp/src/arrow/dataset/file_parquet.cc:494]: The return value needs to be of type Result, so the future needs to be unwrapped. Not sure if this was the right way to do it, but seems to work

**Comment by eeroel** [comment]
Hmm, noticed that the execution freezes when I set io_thread_count <= fragment_readahead. A deadlock probably? I wonder if it's introduced by this change, or if it's a bug that just manifests now because of higher concurrency. I think I have observed such a freeze recently with a recent Pyarrow release, but couldn't reproduce.

File: cpp/src/arrow/dataset/file_base.cc
Comment (bkietz) [reply cpp/src/arrow/dataset/file_base.cc:94]: I think this TODO is valuable but should be linked to a new issue where we extend `custom_open_` to return `Future<std::share_ptr<io::RandomAccessFile>>`. That would be necessary to enable a custom but non blocking open

File: cpp/src/arrow/dataset/file_parquet.cc
Comment (bkietz) [inline cpp/src/arrow/dataset/file_parquet.cc:490]: Instead of copying `input_fut` implicitly by keeping it in closure, I think it'd be more clear to use the value passed to the callback:
```suggestion
      source.OpenAsync().Then([=](std::shared_ptr<arrow::io::RandomAccessFile> input)
                     -> Result<std::unique_ptr<parquet::ParquetFileReader>> {
```

(The `TODO(ARROW-12259)` only applies to futures of move-only types as returned by `ParquetFileReader::OpenAsync()`, and doesn't apply to the return value of `source.OpenAsync()`.)

File: cpp/src/arrow/dataset/file_parquet.cc
Comment (bkietz) [reply cpp/src/arrow/dataset/file_parquet.cc:494]: For a continuation (callback passed to Future::Then) the return type can also be Future. Please change it to Future, then you can return `rfut` directly instead of blocking in a continuation

**Review by bkietz — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/dataset/file_parquet.cc
Comment (eeroel) [reply cpp/src/arrow/dataset/file_parquet.cc:494]: Ah, I see, thanks. For some reason I get a `note: copy constructor is implicitly deleted because 'unique_ptr<parquet::ParquetFileReader>' has a user-declared move constructor` if I try to return the future, any idea what I'm missing?

**Comment by mapleFU** [comment]
> Hmm, noticed that the execution freezes when I set io_thread_count <= fragment_readahead (testing with Pyarrow). A deadlock probably? I wonder if it's introduced by this change, or if it's a bug that just manifests now because of higher concurrency. I think I have observed such a freeze recently with a recent Pyarrow release, but couldn't reproduce.

@eeroel https://github.com/apache/arrow/pull/37918/files#r1340188707 I think this is the problem that we call blocking in async execution.

It's a bit like this problem I fixed: https://github.com/apache/arrow/pull/37514

File: cpp/src/arrow/dataset/file_parquet.cc
Comment (bkietz) [reply cpp/src/arrow/dataset/file_parquet.cc:490]: Sorry, let me give a more complete suggestion

```c++
Future<std::shared_ptr<parquet::arrow::FileReader>> ParquetFileFormat::GetReaderAsync(
    const FileSource& source, const std::shared_ptr<ScanOptions>& options,
    const std::shared_ptr<parquet::FileMetaData>& metadata) const {
  ARROW_ASSIGN_OR_RAISE(
      auto parquet_scan_options,
      GetFragmentScanOptions<ParquetFragmentScanOptions>(kParquetTypeName, options.get(),
                                                         default_fragment_scan_options));
  auto properties =
      MakeReaderProperties(*this, parquet_scan_options.get(), options->pool);

  auto self = checked_pointer_cast<const ParquetFileFormat>(shared_from_this());

  return source.OpenAsync().Then([=](const std::shared_ptr<io::RandomAccessFile>& input) mutable {
    return parquet::ParquetFileReader::OpenAsync(input, std::move(properties), metadata)
        .Then(
            [=](const std::unique_ptr<parquet::ParquetFileReader>& reader) mutable
            -> Result<std::shared_ptr<parquet::arrow::FileReader>> {
              auto arrow_properties = MakeArrowReaderProperties(
                  *self, *reader->metadata(), *options, *parquet_scan_options);

              std::unique_ptr<parquet::arrow::FileReader> arrow_reader;
              RETURN_NOT_OK(parquet::arrow::FileReader::Make(
                  options->pool,
                  // TODO(ARROW-12259): workaround since we have Future<(move-only type)>
                  // It *wouldn't* be safe to const_cast reader except that here we know
                  // there are no other waiters on the reader.
                  std::move(
                      const_cast<std::unique_ptr<parquet::ParquetFileReader>&>(reader)),
                  std::move(arrow_properties), &arrow_reader));

              return std::move(arrow_reader);
            },
            [path = source.path()](const Status& status)
                -> Result<std::shared_ptr<parquet::arrow::FileReader>> {
              return WrapSourceError(status, path);
            });
  });
}
```

Comment (bkietz) [reply cpp/src/arrow/dataset/file_parquet.cc:490]: @lidavidm re my alternative ARROW-12259 workaround

Comment (lidavidm) [reply cpp/src/arrow/dataset/file_parquet.cc:490]: a bit yikes, but if it works

File: cpp/src/arrow/dataset/file_base.cc
Comment (eeroel) [reply cpp/src/arrow/dataset/file_base.cc:94]: Created a new issue and updated the comment https://github.com/apache/arrow/issues/37962

File: cpp/src/arrow/dataset/file_parquet.cc
Comment (eeroel) [reply cpp/src/arrow/dataset/file_parquet.cc]: Thank you very much, updated code to this and it seems to work. The deadlock issue I mentioned earlier is also gone.

**Comment by eeroel** [comment]
> > Hmm, noticed that the execution freezes when I set io_thread_count <= fragment_readahead (testing with Pyarrow). A deadlock probably? I wonder if it's introduced by this change, or if it's a bug that just manifests now because of higher concurrency. I think I have observed such a freeze recently with a recent Pyarrow release, but couldn't reproduce.
> 
> @eeroel https://github.com/apache/arrow/pull/37918/files#r1340188707 I think this is the problem that we call blocking in async execution.
> 
> It's a bit like this problem I fixed: #37514 . The problem is that blocking in async function under same Executor is a bad behavior, which might cause deadlock.

OK, thanks for the explanation. Looks like this is fixed in @bkietz suggestion

**Comment by mapleFU** [comment]
LGTM now. Lets wait for committers review.

**Review by bkietz — APPROVED** [review]

**Comment by mapleFU** [comment]
I think we need to clang-format the file `/arrow/cpp/src/arrow/dataset/file_parquet.cc`, other looks ok to me

**Comment by eeroel** [comment]
> I think we need to clang-format the file `/arrow/cpp/src/arrow/dataset/file_parquet.cc`, other looks ok to me

✅

**Comment by eeroel** [comment]
Sorry, still had an issue with formatting in file_base.cc, should be fixed now

**Comment by mapleFU** [comment]
CI passed, should we merge or trigger more CI @bkietz

**Comment by bkietz** [comment]
@eeroel could you please rebase to pick up the fix ( https://github.com/apache/arrow/pull/37867/files#diff-1bba462ab050e89360fd88110a689e85ee037749cea091a1848ab574381d3795R155 ) for several of the CI failures ( https://github.com/apache/arrow/actions/runs/6383550289/job/17363301059?pr=37918#step:6:1217 )

**Comment by eeroel** [comment]
> @eeroel could you please rebase to pick up the fix ( https://github.com/apache/arrow/pull/37867/files#diff-1bba462ab050e89360fd88110a689e85ee037749cea091a1848ab574381d3795R155 ) for several of the CI failures ( https://github.com/apache/arrow/actions/runs/6383550289/job/17363301059?pr=37918#step:6:1217 )

👍 done

**Comment by bkietz** [comment]
CI failure seems unrelated.

+1, thanks!
I'll merge shortly

**Pull request merged**
